operator/internal/certrotation/options.go

```diff
@@ -0,0 +1,128 @@
+	QuerierHTTPSecret *corev1.Secret
+
+	QuerierGRPC       SelfSignedCertKey
+	QuerierGRPCSecret *corev1.Secret
+
+	QueryFrontendHTTP       SelfSignedCertKey
+	QueryFrontendHTTPSecret *corev1.Secret
+
+	QueryFrontendGRPC       SelfSignedCertKey
+	QueryFrontendGRPCSecret *corev1.Secret
+
+	RulerHTTP       SelfSignedCertKey
+	RulerHTTPSecret *corev1.Secret
+
+	RulerGRPC       SelfSignedCertKey
+	RulerGRPCSecret *corev1.Secret
+}
+
+// SigningCA rotates a self-signed signing CA stored in a secret. It creates a new one when
+// - refresh duration is over
+// - or 80% of validity is over (if RefreshOnlyWhenExpired is false)
+// - or the CA is expired.
+type SigningCA struct {
+	// Validity is the duration from time.Now() until the signing CA expires. If RefreshOnlyWhenExpired
+	// is false, the signing cert is rotated when 80% of validity is reached.
+	Validity time.Duration
+	// Refresh is the duration after signing CA creation when it is rotated at the latest. It is ignored
+	// if RefreshOnlyWhenExpired is true, or if Refresh > Validity.
+	Refresh time.Duration
+	// RefreshOnlyWhenExpired set to true means to ignore 80% of validity and the Refresh duration for rotation,
+	// but only rotate when the signing CA expires. This is useful for auto-recovery when we want to enforce
+	// rotation on expiration only, but not interfere with the ordinary rotation controller.
+	RefreshOnlyWhenExpired bool
+	// RawCA is the CA certificates in bytes extracted from the secret
+	RawCA *crypto.CA
+}
+
+// SelfSignedCertKey rotates a key and cert signed by a signing CA and stores it in a secret.
+//
+// It creates a new one when
+// - refresh duration is over
+// - or 80% of validity is over (if RefreshOnlyWhenExpired is false)
+// - or the cert is expired.
+// - or the signing CA changes.
+type SelfSignedCertKey struct {
+	// Validity is the duration from time.Now() until the certificate expires. If RefreshOnlyWhenExpired
+	// is false, the key and certificate is rotated when 80% of validity is reached.
+	Validity time.Duration
+	// Refresh is the duration after certificate creation when it is rotated at the latest. It is ignored
+	// if RefreshOnlyWhenExpired is true, or if Refresh > Validity.
+	// Refresh is ignored until the signing CA at least 10% in its life-time to ensure it is deployed
+	// through-out the cluster.
+	Refresh time.Duration
+	// RefreshOnlyWhenExpired allows rotating only certs that are already expired. (for autorecovery)
+	// If false (regular flow) it rotates at the refresh interval but no later then 4/5 of the cert lifetime.
+	// RefreshOnlyWhenExpired set to true means to ignore 80% of validity and the Refresh duration for rotation,
+	// but only rotate when the certificate expires. This is useful for auto-recovery when we want to enforce
+	// rotation on expiration only, but not interfere with the ordinary rotation controller.
+	RefreshOnlyWhenExpired bool
+	// CertCreator does the actual cert generation.
+	CertCreator targetCertCreator
```

Comment:
Public attribute of a public struct having a private type is not a good idea. Usually means that either the attribute or the whole struct could be private.